### PR TITLE
Dx 338 multi trait view

### PR DESCRIPTION
**Pull request by rlperez**

### CODE REVIEW

**Comment by rlperez** [comment]
Some of these new classes may need their names reviewed. Not sure they really make sense but that's easy enough to update.

File: lib/providers/device_provider.dart [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      (_state == WidgetState.loading || _state == WidgetState.performingAction);
+
+  /// To know if this ChangeNotifier had an error
+  /// see [getErrorMessage] to get the accomponying error message
+  bool get isInErrorState => _state == WidgetState.error;
+
+  /// Get the error message whenever this ChangeNotifier had an error
+  /// See also: [isInErrorState]
+  String get getErrorMessage => _latestErrorMsg;
+}
```
Comment (surajadh) [inline lib/providers/device_provider.dart:101]: Thanks for cleaning up this code and creating a superclass for this, looks cleaner

File: lib/providers/trait_detail_provider.dart [hunk 2]
```diff
@@ -22,15 +22,13 @@ class TraitDetailProvider extends ChangeNotifier {
     fetchData(getDetails: getDetails);
   }
 
-  Future<Device?> fetchData(
+  Future<void> fetchData(
```
Comment (surajadh) [inline lib/providers/trait_detail_provider.dart:25]: Please ensure if this was used just as others as well. Looks like changeNotifier is not called here as well.

File: lib/traits/detail_screen.dart [hunk 3]
```diff
@@ -51,7 +54,7 @@ class DetailScreenWidget extends StatelessWidget {
     return Consumer<TraitDetailProvider>(
         builder: (_, traitBasedDeviceNotifier, child) {
       if (traitBasedDeviceNotifier.isLoading) {
-        return CircularProgressIndicator();
+        return Center(child: CircularProgressIndicator());
```
Comment (surajadh) [inline lib/traits/detail_screen.dart:57]: ~~Not sure why this change is needed, was there some visual change that makes this better?~~

File: lib/traits/detail_screen.dart [hunk 4]
```diff
@@ -63,112 +66,84 @@ class DetailScreenWidget extends StatelessWidget {
...
             child: Column(children: [
-          Row(children: [
-            SizedBox(width: 50),
-            createTraitWidget(traits.first),
-          ]),
-          ...traits
-              .skip(1)
-              .map((trait) => createSlimTraitWidget(trait))
-              .toList(),
+          _card(_createTraitWidget(traits.first), const EdgeInsets.all(8.0),
```
Comment (surajadh) [inline lib/traits/detail_screen.dart:69]: ~~Same here not clear why this refactoring is needed.~~

Comment (surajadh) [reply lib/traits/detail_screen.dart:69]: OK I think I understand it now. Disregard the previous comment.

File: lib/traits/detail_screen.dart [hunk 5]
```diff
@@ -63,112 +66,84 @@ class DetailScreenWidget extends StatelessWidget {
...
-  }
-
-  Color _getBatteryLevelColor(int batteryLevel) {
-    if (batteryLevel >= WidgetStyleConstants.batteryFullMin) {
-      return WidgetStyleConstants.globalSuccessColor;
-    } else if (batteryLevel <= WidgetStyleConstants.batteryLowMax) {
-      return WidgetStyleConstants.globalWarningColor;
-    } else {
-      return Colors.white;
+        return DeviceSlimWidget(
```
Comment (surajadh) [inline lib/traits/detail_screen.dart:137]: Is this UnknownDeviceSlimWidget

File: lib/traits/lock_widget.dart [hunk 6]
```diff
@@ -54,6 +71,6 @@ class LockWidget extends StatelessWidget {
 
   void _lockTap(LockProvider provider) {
     bool setLock = !provider.isLocked;
-    provider.setLockUnlockAction(provider.deviceDetail.id, setLock);
+    provider.setLockUnlockAction(provider.deviceDetail?.id ?? '', setLock);
```
Comment (surajadh) [inline lib/traits/lock_widget.dart:74]: Can the action be called with empty parameter. This doesn't seem right setup.

File: lib/traits/slim/device_slim_widget.dart [hunk 7]
```diff
@@ -0,0 +1,64 @@
+import 'package:flutter/material.dart';
```
Comment (surajadh) [inline lib/traits/slim/device_slim_widget.dart:1]: Is this UnknownDevice?

**Review by surajadh — CHANGES_REQUESTED** [review]

File: lib/components/lock_widget.dart [hunk 8]
```diff
@@ -50,7 +50,7 @@ class LockWidget extends StatelessWidget {
   }
 
   Widget getLockStateIcon(LockProvider lockProvider) {
-    return (lockProvider.loadingDetail || lockProvider.loadingAction)
+    return (lockProvider.isLoading || lockProvider.isPerformingAction)
```
Comment (ses110) [inline lib/components/lock_widget.dart:53]: Could use lockProvider.isBusy instead of having to check for either isLoading or isPerformingAction

File: lib/providers/device_provider.dart (hunk 1 above)
Comment (ses110) [reply lib/providers/device_provider.dart:101]: Agreed, this was a great and much needed refactor

File: lib/traits/slim/device_slim_widget.dart (hunk 7 above)
Comment (rlperez) [reply lib/traits/slim/device_slim_widget.dart:1]: This is more of a generic base so whenever we need something new it's just plug and play whatever icons and content.

Comment (surajadh) [reply lib/traits/slim/device_slim_widget.dart:1]: Ok, I see. I think this maybe one of those instances where it makes intention clearer to name this as a baseDevice.

File: lib/traits/lock_widget.dart (hunk 6 above)
Comment (rlperez) [reply lib/traits/lock_widget.dart:74]: It appears to create some sort of closure because I thought it would be odd too but verified with the debugger that the device id makes it into the repository call.

**Review by surajadh — APPROVED** [review]
